docs/dev-process/coding-guidelines.md

```diff
@@ -1,92 +1,179 @@
+* Update/remove existing comments when you are changing the corresponding code.
 
-The PowerShell repo is used to build PowerShell targeting CoreCLR as well as CLR 4.5.
+* Make sure the added/updated comments are accurate and easy to understand.
 
-Code under !CORECLR must build against CLR 4.5.
-We will not accept changes that require a later version of the full CLR.
-In extremely rare cases, we may use reflection to use an API in a later version of the CLR,
-but the feature must robustly handle running with CLR 4.5.
+* Public members must use [doc comments](https://msdn.microsoft.com/en-us/library/b2s063f7.aspx).
+  Internal and private members may use doc comments but it is not required.
 
-We may reject code under !CORECLR without explanation because
-we do not support installation or testing of such code in this repo.
-All new features should support CoreCLR.
-
-## Performance considerations
+## Performance Considerations
 
 PowerShell has a lot of performance sensitive code as well as a lot of inefficient code.
-We have some guidelines that we typically apply widely even in less important code
-because code and patterns are copied we want certain inefficient code to stay out of the performance critical code.
+We have some guidelines that we typically apply widely even in less important code because code and patterns are copied,
+and we want certain inefficient code to stay out of the performance critical code.
 
 Some general guidelines:
 
-* Avoid LINQ - it can create lots of avoidable garbage
+* Avoid LINQ - it can create lots of avoidable garbage.
+
 * Prefer `for` and `foreach`,
   with a slight preference towards `for` when you're uncertain if `foreach` allocates an iterator.
+
 * Avoid `params` arrays, prefer adding overloads with 1, 2, 3, and maybe more parameters.
+
 * Be aware of APIs such as `String.Split(params char[])` that do not provide overloads to avoid array allocation.
-  When calling such APIs, reuse a static array when possible.
+  When calling such APIs, reuse a static array when possible (i.e. `Utils.Separators.Colon`).
+
+* Avoid creating empty arrays.
+  Instead, reuse the static ones via `Utils.EmptyArray<T>`.
+
 * Avoid unnecessary memory allocation in a loop.
   Move the memory allocation outside the loop if possible.
 
-## Portable code
+* Avoid gratuitous exceptions as much as possible.
+  Exception handling can be expensive due to cache misses and page faults when accessing the handling code and data.
+  Finding and designing away exception-heavy code can result in a decent performance win.
+  For example, you should stay away from things like using exceptions for control flow.
+
+* Use `dict.TryGetValue` instead of `dict.Contains` and `dict[<key>]` when retrieving value from a `Dictionary`.
+  In this way you can avoid hashing the key twice.
+
+## Security Considerations
+
+Security is an important aspect of PowerShell and we need to be very careful about changes that may introduce security risks.
```

Comment:
I feel we should say a little more about what security risks are and should at least mention input validation, code injection, privilege escalation, data privacy, and DOS concerns.  We should have a separate document that goes into more detail but that can be a separate effort.

Reply:
`input validation` - this could be easily confused with the input validation within an API. Could you please elaborate a bit on it in the security context?

`data privacy` - do you mean data privacy breach? Can you please give an example?

> We should have a separate document that goes into more detail

That would be very useful. Can you please work on such a doc? If you are busy on things and cannot get to it in near future, I'm happy to work on a draft if you can provide me some resources and pointers.

Reply:
Has CoreFX repo such document?

Reply:
Input validation for public API or Cmdlet is exactly what I mean :).   That is where code injection can occur and I feel being aware of it should be part of all code reviews.  Likewise data privacy is being aware that private data must always be protected.  It is not something we have to be too concerned with with PowerShell (other than credentials and transferring remote data) but again something to be aware of during a code review.

I think we should just mention basic security areas to be aware of during a code review in a general way here, and then provide a link to a separate document where we can provide more details and examples.  I am happy to work on such a document but I am not sure when I can get to it.  I'll create an issue to track the work.

Reply:
#4759

Reply:
Add the following:
```
... may introduce security risks,
such as code injection caused by the lack of input validation,
privilege escalation due to the misuse of impersonation,
or data privacy breach with a plain text password.
```